### PR TITLE
yubikey-agent: init at 0.1.3

### PR DESCRIPTION
**Pull request by philandstuff**
###### Motivation for this change

This adds yubikey-agent as a package and a nixos module.

The yubikey-agent is impure in how it calls pinentry; I tried patching
gopass to point at a particular `pinentry` but this changes the value of
`vendorSha256` depending on the particular choice of pinentry program,
which felt frustratingly difficult.  Instead, we use `wrapProgram` to
set a default pinentry in PATH, and use a systemd override to prefix
PATH to select a chosen pinentry program if specified.

On linux, we need libnotify to provide the notify-send utility for
desktop notifications (such as "Waiting for Yubikey touch...").

The systemd unit a copy of the one in the upstream repo, in systemd.md:
https://github.com/FiloSottile/yubikey-agent/blob/master/systemd.md

This might work on other flavors of unix, but I haven't tested.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (flokli) [inline pkgs/tools/security/yubikey-agent/default.nix:22]: Please link to https://github.com/go-piv/piv-go/pull/75 here, so we know this will eventually get fixed and can be removed.

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (flokli) [inline pkgs/tools/security/yubikey-agent/default.nix:1]: We shouldn't need to pass in any pinentry here.

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (flokli) [inline pkgs/tools/security/yubikey-agent/default.nix:3]: We shouldn't need to construct any pinentryBinaryPath here.

**Review by flokli — CHANGES_REQUESTED** [review]
Thanks for the code! I don't think we should reinvent yet another pinentry selection, so we should just reuse the one from the gnupg module for the time being.

File: nixos/modules/services/security/yubikey-agent.nix
Comment (flokli) [reply nixos/modules/services/security/yubikey-agent.nix:77]: You can also just set `systemd.user.services.<name>.path` to add the selected pinentry to `$PATH`.

**Comment by flokli** [comment]
Is there any context/docs on why `yubikey-agent` is fine to discover `pinentry` from `$PATH`, whereas `gpg-agent.conf` insists on absoulte paths?

I did some digging in https://github.com/NixOS/nixpkgs/pull/90214#issuecomment-652487240, so would be nice to hear @FiloSottile's opinion on that as well :-)

File: pkgs/tools/security/yubikey-agent/yubikey-agent.service
Comment (flokli) [inline pkgs/tools/security/yubikey-agent/yubikey-agent.service:1]: This could also just end up being a template in https://github.com/FiloSottile/yubikey-agent.git. Wanna file a PR?

We might also want to add a `yubikey-agent.socket` file, to start the unit on demand when  `%t/yubikey-agent/yubikey-agent.sock` is accessed.

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (philandstuff) [reply pkgs/tools/security/yubikey-agent/default.nix:22]: :+1: I may even just patch it to use go-piv/piv-go#75 directly, using a `go.mod` `replace` directive, because it's a better solution.

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (philandstuff) [reply pkgs/tools/security/yubikey-agent/default.nix:41]: Yes it is, for users who don't run yubikey-agent via systemd.  In particular, the systemd unit is no use on macOS, but yubikey-agent needs a way to find pinentry-mac.

**Comment by philandstuff** [comment]
@flokli thank you for your thorough review! I agree with most of your points and will update the PR.

Yes, the way `pinentry` is configured is weird and idiosyncratic.

The gnupg nix package sets a hardcoded absolute path to pinentry.  This works ok because programs like `gpg-agent` have a `--pinentry-program` command-line option to choose a different pinentry at runtime, as well as config file options and suchlike. This means that the `gnupg` nixos module can allow the user to configure their choice of pinentry program easily, even if the underlying `gnupg` package has a hardcoded pinentry path as a default value.

yubikey-agent pulls pinentry from PATH, and it does so in a 3rd party module ([it's actually within `gopass`](https://pkg.go.dev/github.com/gopasspw/gopass/pkg/pinentry?tab=doc)), which [makes it incredibly awkward to patch an absolute path to pinentry instead](https://discourse.nixos.org/t/patching-dependencies-in-go-modules/7972).  yubikey-agent also offers no way to choose a pinentry program at runtime other than by putting the desired one first in PATH.  It has some code to dispatch to a different binary name on different OSes (`pinentry.exe` on windows, `pinentry-mac` on macOS, `pinentry` otherwise) but on non-macOS unixes it basically assumes the underlying distribution will provide an appropriate `pinentry`, in the PATH.  In fairness this is the case with most distributions, but with nix it's completely against the philosophy.

I'm not sure I can think of a better alternative.

File: nixos/modules/services/security/yubikey-agent.nix
Comment (philandstuff) [reply nixos/modules/services/security/yubikey-agent.nix:75]: I tried doing this, but I got:

```
error: The option `programs.gnupg.agent.pinentryFlavor' is used but not defined.
(use '--show-trace' to show detailed location information)
```

I'm not sure what I'm doing wrong here, or if there's something clever I need to do to depend on another module's configuration.

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (philandstuff) [reply pkgs/tools/security/yubikey-agent/default.nix:45]: @flokli I'm happy to do this, but can you help me learn what the benefit is of doing it this way?

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (flokli) [reply pkgs/tools/security/yubikey-agent/default.nix:22]: Still you might want to add the note, so people know when to remove it again.

**Comment by flokli** [comment]
I mean, we might want to open a discussion with the GnuPG people about why pinentry discovery there is the way it is, and why it's not defaulting to looking for it in `$PATH` if it's neither set in `gpg-agent.conf`, nor in `--pinentry-program`.

Discovering from `$PATH` for gpg as well might help (https://github.com/NixOS/nixpkgs/pull/90214#issuecomment-645270366), but it'd like to understand the reasoning on why it currently doesn't do that (and if `yubikey-agent`/`pass` doing differently is actually a footgun - if it's not, I'm all for adding it to its `$PATH` as suggested in https://github.com/NixOS/nixpkgs/pull/92936#discussion_r453308280).

**Comment by philandstuff** [comment]
Let’s assume we go with your proposal to just look for pinentry in PATH (for both gnupg and yubikey-agent). What happens if there is no `pinentry` anywhere on the PATH?

**Comment by FiloSottile** [comment]
`yubikey-agent` intentionally doesn't have a config file. Would a `PINENTRY` env var work for you? Happy to add it.

**Comment by philandstuff** [comment]
Okay, I've pushed a couple of updates.  Key unresolved issues:

 - [ ] I agree that I should reuse the `gnupg.agent.pinentryFlavor` but [i get an error](https://github.com/NixOS/nixpkgs/pull/92936#discussion_r453326936)
 - [ ] We need to decide how, and where, to configure pinentry

To expand on the pinentry discussion: as a starting point, I feel that a nix closure should contain everything it needs to function correctly.  This is why, in general, nix packages depend on other packages via absolute store paths and not via PATH.  If you have a dependency via PATH, that does not form part of the closure, and when you copy the closure to another system, you won't copy the dependency.

However, I can see how pinentry is a bit of a weird edge case: users might reasonably want to configure different types of pinentry, and they shouldn't have to install all the pinentries in order to use just one of them.

It's also worth noting that the dynamics on macOS are different: on macOS, there is only one pinentry implementation, and I think gnupg and yubikey-agent should just use it without any extra configuration.  But on Linux (and probably other unixes) it's useful to be able to dynamically choose a pinentry program.

I must confess I read some of the discussions that @flokli linked to above (such as https://github.com/NixOS/nixpkgs/pull/92936#discussion_r453308280) but it's really hard to get context by jumping partway through a conversation and chasing back through many linked issues so it's mostly going above my head.  But my understanding is: 1) gnupg has a baked-in pinentry choice, 2) gnupg allows users to choose a different pinentry at invocation time, 3) there is some desire (I don't know why) that gnupg should be able to use pinentry from PATH.  In any case, the design of gnupg sounds like a live issue, so we shouldn't assume the current gnupg packaging is the correct way to go.

Right now my feeling is:

 - on macOS, the built yubikey-agent binary should use pinentry-mac without any extra configuration
 - on Linux, the built yubikey-agent could just get whatever pinentry it finds in PATH, and we don't need to bake in a default.
 - if the gnupg side of the discussion is resolved, I'm happy to try to ensure yubikey-agent is consistent with it.

How does that sound?

**Comment by philandstuff** [comment]
@FiloSottile Thanks for the offer! I think we need to discuss things a bit further here but we may well take you up on this.

File: nixos/modules/services/security/yubikey-agent.nix
Comment (flokli) [reply nixos/modules/services/security/yubikey-agent.nix:75]: If works if you don't assign the whole `config` attrset at once:

```diff
index d378666b4f1..4c07b9227d6 100644
--- a/nixos/modules/services/security/yubikey-agent.nix
+++ b/nixos/modules/services/security/yubikey-agent.nix
@@ -7,19 +7,6 @@ with lib;
 let
   cfg = config.services.yubikey-agent;
 
-  # Copied from the gnupg module
-  xserverCfg = config.services.xserver;
-  defaultPinentryFlavor =
-    if xserverCfg.desktopManager.lxqt.enable
-    || xserverCfg.desktopManager.plasma5.enable then
-      "qt"
-    else if xserverCfg.desktopManager.xfce.enable then
-      "gtk2"
-    else if xserverCfg.enable || config.programs.sway.enable then
-      "gnome3"
-    else
-      null;
-
 in
 {
   ###### interface
@@ -46,41 +33,26 @@ in
           The package used for the yubikey-agent daemon.
         '';
       };
-
-      pinentryFlavor = mkOption {
-        type = types.nullOr (types.enum pkgs.pinentry.flavors);
-        example = "gnome3";
-        description = ''
-        Which pinentry interface to use. If not null, the path to the
-        pinentry binary will be provided to yubikey-agent by setting the PATH
-        appropriately in the systemd service and thus overrides the pinentry
-        option in the yubikey-agent package's default systemd unit.
-        If not set at all, it'll pick an appropriate flavor depending on the
-        system configuration (qt flavor for lxqt and plasma5, gtk2 for xfce
-        4.12, gnome3 on all other systems with X enabled, ncurses otherwise).
-      '';
-      };
-
     };
   };
 
-  config = {
-    services.yubikey-agent.pinentryFlavor = mkDefault defaultPinentryFlavor;
 
-    environment.systemPackages = [ cfg.package ];
-    systemd.packages = [ cfg.package ];
+  config.environment = {
+    extraInit = optionalString cfg.enable ''
+      if [ -z "$SSH_AUTH_SOCK" -a -n "$XDG_RUNTIME_DIR" ]; then
+        export SSH_AUTH_SOCK="$XDG_RUNTIME_DIR/yubikey-agent/yubikey-agent.sock"
+      fi
+    '';
+    systemPackages = [ cfg.package ];
+  };
+
+  config.systemd = {
+    packages = [ cfg.package ];
 
     # This overrides the systemd user unit shipped with the
     # yubikey-agent package
-    systemd.user.services.yubikey-agent = mkIf (cfg.pinentryFlavor != null) {
-      path = [ pkgs.pinentry.${cfg.pinentryFlavor} ];
+    user.services.yubikey-agent = mkIf (config.programs.gnupg.agent.pinentryFlavor != null) {
+      path = [ pkgs.pinentry.${config.programs.gnupg.agent.pinentryFlavor} ];
     };
-
-    environment.extraInit = optionalString cfg.enable
-      ''
-        if [ -z "$SSH_AUTH_SOCK" -a -n "$XDG_RUNTIME_DIR" ]; then
-          export SSH_AUTH_SOCK="$XDG_RUNTIME_DIR/yubikey-agent/yubikey-agent.sock"
-        fi
-      '';
   };
 }
```

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (flokli) [reply pkgs/tools/security/yubikey-agent/default.nix:35]: At least on Linux, `yubikey-agent` could set `PATH` in the `.service` file, so it should already find the right pinentry.

From the screenshots, `yubikey-agent -setup` doesn't look like it uses pinentry to ask for the setup pin, and the README also only suggests starting it via systemd, so I don't really see a need to wrap it with a pinentry (which here would also be `pinentry-curses` btw and probably not what we want).

Even when people use this on non-NixOS, they should have their configured `pinentry` in `$PATH`.

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (flokli) [reply pkgs/tools/security/yubikey-agent/default.nix:39]: its cleaner, and the somewhat recommended way to patch these things. If the `.service` suddenly switches from `/usr/local/bin` to `/bin`, we just end up not substituting anymore, and the unit silently breaks.

The patch will fail loudly.

File: pkgs/tools/security/yubikey-agent/yubikey-agent.service
Comment (flokli) [reply pkgs/tools/security/yubikey-agent/yubikey-agent.service:1]: I opened a PR at https://github.com/FiloSottile/yubikey-agent/pull/43.

Comment (philandstuff) [reply pkgs/tools/security/yubikey-agent/yubikey-agent.service:1]: I'm quite new to systemd, but I'm not sure that `yubikey-agent.socket` will work for us?  yubikey-agent doesn't listen on stdin, it listens on a named file, which doesn't seem compatible with allowing systemd to pass the socket in to yubikey-agent?  I've probably misunderstood something though.

Comment (flokli) [reply pkgs/tools/security/yubikey-agent/yubikey-agent.service:1]: Erp, I assumed this already was able to use sockets passed in from systemd (https://vincent.bernat.ch/en/blog/2018-systemd-golang-socket-activation). I'll remove the `.socket` bits from the PR, thanks!

File: nixos/modules/services/security/yubikey-agent.nix
Comment (philandstuff) [reply nixos/modules/services/security/yubikey-agent.nix:75]: I applied this patch, exactly as in your message, on top of d883cfad, but I still got the same problem.

I've found the issue: my configuation.nix doesn't set `programs.gnupg.agent.enable` or `programs.gnupg.agent.pinentryFlavor`.  The gnupg mkDefault is only evaluated if `programs.gnupg.agent.enable` is true; which means you can only get the useful defaulting behaviour if you choose to run the gpg-agent.

It would be possible, but not easy, to edit the gnupg module to always apply the `mkDefault`, but I'm not sure it's the right solution for us here.

I don't like the current copypaste solution but I don't think we've found anything better quite yet. Maybe we could extract a separate `pinentry` module, which just does the defaulting behaviour, and reference it from `gnupg` and `yubikey-agent`?

File: pkgs/tools/security/yubikey-agent/default.nix
Comment (philandstuff) [reply pkgs/tools/security/yubikey-agent/default.nix:39]: I suspect that moving to `substituteAll` and moving to an upstream systemd service like in https://github.com/FiloSottile/yubikey-agent/pull/43 are mutually exclusive.  I'm more keen on using an upstream systemd service, so I'll see how that PR plays out.

**Comment by philandstuff** [comment]
Added @rawkode as a maintainer, ref: https://github.com/NixOS/nixpkgs/pull/92965#issuecomment-657221939

**Comment by philandstuff** [comment]
I pushed another update.  I've pulled in the systemd unit from https://github.com/FiloSottile/yubikey-agent/pull/43 but the relative path trick gave me the error: `Executable "yubikey-agent" not found in path "/nix/store/zzr1p57r1ii6y0gm7kycid0gkw1hskp3-systemd-245.6/bin/"` so I still need to substitute the nix store path into it.

I have removed baking in pinentry to PATH, except on macOS where we bake in pinentry_mac since there's no other choice.

I still haven't come up with a solution to reusing the gpg-agent pinentry selection (see https://github.com/NixOS/nixpkgs/pull/92936#discussion_r454628056).

**Comment by flokli** [comment]
> I pushed another update. I've pulled in the systemd unit from [FiloSottile/yubikey-agent#43](https://github.com/FiloSottile/yubikey-agent/pull/43) but the relative path trick gave me the error: `Executable "yubikey-agent" not found in path "/nix/store/zzr1p57r1ii6y0gm7kycid0gkw1hskp3-systemd-245.6/bin/"` so I still need to substitute the nix store path into it.

Yeah, this is expected.`systemd` luckily doesn't pick up things from `$PATH` when setting just a binary name in `ExecStart=` but looks up from paths baked into systemd.

On NixOS, we [set](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/systemd/0018-path-util.h-add-placeholder-for-DEFAULT_PATH_NORMAL.patch) these to `[$systemd/bin](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/systemd/default.nix#L68)` only for reproducibility.

> I have removed baking in pinentry to PATH, except on macOS where we bake in pinentry_mac since there's no other choice.

Is there not a native pinentry provided in `$PATH`?

> I still haven't come up with a solution to reusing the gpg-agent pinentry selection (see [#92936 (comment)](https://github.com/NixOS/nixpkgs/pull/92936#discussion_r454628056)).

Yeah, this is still somewhat pending. I'd probably prefer moving pinentry selection into a module both used from the gnupg module and this module over duplicating it…

**Comment by philandstuff** [comment]
> Is there not a native pinentry provided in $PATH?

macOS doesn’t provide one, no. The user might have installed one via gpgtools or homebrew, but plain macOS doesn’t have a pinentry.

**Comment by philandstuff** [comment]
> Is there not a native pinentry provided in $PATH?

macOS doesn’t provide one, no. The user might have installed one via gpgtools or homebrew, but plain macOS doesn’t have a pinentry.

**Comment by philandstuff** [comment]
@flokli I've just pushed an update.  It turns out the magic word I needed was `mkMerge` https://nixos.org/nixos/manual/#sec-option-definitions-merging

I think this resolves all the issues we had.  PTAL :smile:

**Comment by rawkode** [comment]
I'm learning a lot from monitoring this, thank you for your persistence @philandstuff

File: nixos/modules/programs/gnupg.nix
Comment (flokli) [inline nixos/modules/programs/gnupg.nix:96]: you can just set a `default = defaultPinentryFlavor` in the options list above (https://github.com/NixOS/nixpkgs/pull/92936/files#diff-46c0e5e66d829e49bf18bc90891c2c90L72), and not use the `mkMerge` and the reformatting.

**Review by flokli — CHANGES_REQUESTED** [review]
Thanks for the persistence too :-) I think the pinentry problem could be solved without the `mkMerge`, sent another review.

File: nixos/modules/programs/gnupg.nix
Comment (philandstuff) [reply nixos/modules/programs/gnupg.nix:96]: Huh, I just assumed there was some reason it wasn’t done that way originally (to avoid infinite recursion loops?) but if that works I can just do that. I’m fairly new to nixos modules and options and they scare me a bit 😰

**Comment by philandstuff** [comment]
@flokli pushed another update

**Comment by flokli** [comment]
This LGTM. I poked upstream for a new release containing the unit file in https://github.com/FiloSottile/yubikey-agent/pull/43#issuecomment-661380666, so we could avoid duplicating this.

cc @Ma27 on the pinentry changes.

**Comment by flokli** [comment]
Okay, let's merge this as-is. We can always move to the upstream-provided systemd unit, once there's a new release. :+1: 

Thanks for your work!

**Comment by vcunat** [comment]
Another problematic platform is i686.  I expect it would be OK not to build on unusual platforms **unless** it made default NixOS depend on the package.

I checked it's not fixed by #93850.  You test on x86_64 by building `pkgsi686Linux.yubikey-agent`.  [Example failure on Hydra](https://hydra.nixos.org/build/124329863).  Almost all i686 NixOS tests are broken now.

**Comment by flokli** [comment]
@vcunat Uh, somehow I missed this was missing a `mkIf cfg.enable`, pulling in yubikey-agent unconditionally :facepalm: 

PR at https://github.com/NixOS/nixpkgs/pull/93892.

File: nixos/modules/services/security/yubikey-agent.nix
Comment (philandstuff) [inline nixos/modules/services/security/yubikey-agent.nix:45]: @vcunat oops! Thanks for the report. I guess a quick fix would be to only add yubikey-agent to `environment.systemPackages` and `systemd.packages` conditionally on `cfg.enable` (which we should have done anyway).

Longer term, it’d be better to make it build on i686 but I think unblocking the build is more important right now. Unfortunately I’m afk for the next few hours so can’t immediately raise a PR to fix.

Comment (flokli) [reply nixos/modules/services/security/yubikey-agent.nix:45]: I already raised a PR - see my comment above.

**Comment by flokli** [comment]
The PR was merged, so we can try to fix i686 at a later time. I opened https://github.com/go-piv/piv-go/issues/76 for investigation.

**Pull request merged**
